rdrf/rdrf/helpers/utils.py

```diff
@@ -705,6 +707,26 @@ def get_supported_languages():
     return [Language(pair[0], pair[1]) for pair in settings.LANGUAGES]
 
 
+LanguageInfo = namedtuple('Language', ['code', 'name'])
+
+
+def get_all_language_codes():
+    languages_in_settings = dict(settings.LANGUAGES)
+    language_codes = set(languages_in_settings.keys())
+    extra_language_codes = [language_code for language_code in LANGUAGE_ALPHA3.keys() if not any(language_code in code for code in language_codes)]
+    language_codes.update(extra_language_codes)
```

Comment:
This is redundant because adding items to the set will check that they aren't already in there.

```suggestion
    language_codes.update(LANGUAGE_ALPHA3.keys())
```

Reply:
This is trying to avoid having the same language with different language code. E.g. In django settings, we have `zh-CN` and `zh-TW` representing for '_Simplify Chinese_' and '_Traditional Chinese_' respectively, while in LANGUAGE_ALPHA3, `zh` is used to represent for '_Chinese_' as a whole. Without using code `extra_language_codes = [language_code for language_code in LANGUAGE_ALPHA3.keys() if not any(language_code in code for code in language_codes)]` , `zh, zh-CN, zh-TW` will be all in the preferred language list. Similarly, ar and ar-AR are duplicated as well.

Reply:
I see :+1: 

This appears to be a broader problem relating to [macrolanguages](https://en.wikipedia.org/wiki/ISO_639_macrolanguage), which aren't strictly speaking 'languages' themselves.

It seems that to exclude macrolanguages using langcodes, you would use something like `set(NORMALIZED_MACROLANGUAGES.values())`, however this set would exclude languages that are technically macrolanguages, but on the web seem to refer to a more standardized single language (eg. Arabic, Estonian, Nepali).

I think it doesn't really make sense for someone to register as speaking 'Chinese' (`zh`), so we definitely shouldn't let people register with that as a preferred language.

So I would say the problem is that we're using the `LANGUAGE_ALPHA3` dict as a list of languages that people use to read text on the web, which I'm not sure is what it should be used for.

Instead, we could just use the list of languages that Django has support for (`django.conf.locale.LANG_INFO`): 

https://github.com/django/django/blob/77278929c86168f075600d9d8c8e76a4792e672b/django/conf/locale/__init__.py#L10-L27

The only problem I can see is that this would also mean that users can't register with a preferred language that we can't support (eg. Malay until the Django 4 update), because Django limits the supported languages to those in that list.

What do you think @Lexiel46 and @ppettitau?

Reply:
From my understanding, the aim of the ticket is to invite users to contribute to the crowdin translations when accessesing the registry in a language that is not supported. Hence, I'm afraid it may be against the original intention if users can't register with a preferred language that we can't support.

There're some other dicts with language codes in `data_dicts.py` in `langcodes` but I didn't find a more proper one than `LANGUAGE_ALPHA3`.

Reply:
Use this as a starting point:

```python
languages_in_settings = dict(settings.ALL_LANGUAGES)
language_codes = set(languages_in_settings.keys())
language_codes_simple = set(s.split("-")[0] for s in language_codes)
extra_language_codes = set(language_code for language_code in LANGUAGE_ALPHA3.keys()).difference(language_codes_simple)
final_language_codes = language_codes.union(extra_language_codes)
```

Reply:
Indonesian appears twice because the `id` langauge code has replaced the `in` langauge code, but the latter is still included in the dataset. It is marked as deprecated, but still appears in the list of codes: https://github.com/rspeer/langcodes/blob/49beea8e20ae26c2dead7bd77f41cfba0e0ab533/langcodes/data/language-subtag-registry.txt#L428

It appears you're supposed to us `standardize_tag` to fix these deprecated values: https://github.com/rspeer/langcodes/tree/master#standardizing-language-tags